src/compiler/transformers/module/module.ts

```diff
@@ -482,12 +482,95 @@ namespace ts {
                     return visitEndOfDeclarationMarker(<EndOfDeclarationMarker>node);
 
                 default:
-                    // This visitor does not descend into the tree, as export/import statements
-                    // are only transformed at the top level of a file.
-                    return node;
+                    return visitEachChild(node, importCallExpressionVisitor, context);
             }
         }
 
+        function importCallExpressionVisitor(node: Node): VisitResult<Node> {
+            // This visitor does not need to descend into the tree if there is no dynamic import,
+            // as export/import statements are only transformed at the top level of a file.
+            if (!(node.transformFlags & TransformFlags.ContainsDynamicImport)) {
+                return node;
+            }
+
+            switch (node.kind) {
+                case SyntaxKind.ImportCallExpression:
+                    return visitImportCallExpression(<ImportCallExpression>node);
+                default:
+                    return visitEachChild(node, importCallExpressionVisitor, context);
+            }
+        }
+
+        function visitImportCallExpression(node: ImportCallExpression): Expression {
+            switch (compilerOptions.module) {
+                case ModuleKind.CommonJS:
+                    return transformImportCallExpressionCommonJS(node);
+                case ModuleKind.AMD:
+                    return transformImportCallExpressionAMD(node);
+                case ModuleKind.UMD:
+                    return transformImportCallExpressionUMD(node);
+            }
+            Debug.fail("All supported module kind in this transformation step should have been handled");
+        }
+
+        function transformImportCallExpressionUMD(node: ImportCallExpression): Expression {
+            // (function (factory) {
+            //      ... (regular UMD)
+            // }
+            // })(function (require, exports, useSyncRequire) {
+            //      "use strict";
+            //      Object.defineProperty(exports, "__esModule", { value: true });
+            //      require.length === 1 ?
+            //          /*CommonJs Require*/ Promise.resolve().then(() => require('blah'));
+            //          /*Amd Require*/ new Promise(resolve => require(['blah'], resolve));

```

Comment:
Can we be sure that every AMD and CJS implementation will have a `require` with a reliable `length` property? I'd rather do proper feature detection to be safe. 

Also, tools that do static analysis of `require(id)` from the output javascript (such as the System loader) will have issues with us emitting a `require` call without a valid module id. I would recommend we emit the following scoped helpers at the top of the module body function:

```ts
var __require = require;
var __import = typeof module === "object" && typeof module.exports === "object" 
  ? function (id) { return new Promise(function (resolve) { resolve(); }).then(function () { return __require(id); }); }
  : function (id) { return new Promise(function (resolve, reject) { __require([id], resolve, reject); }); };
```

This has the following benefits:

1. Less repetition of promise creation logic (the __import helper is defined once).
2. No dependency on `Promise.resolve` which is not Promise/A+ compliant (though it is valid in the ES6 spec).
3. Aliasing `require` as `__require` should hopefully avoid static analysis issues with the System loader, WebPack, browserify, etc.
4. Passes the `reject` callback to the AMD `require` call. Though not officially specified, many AMD loaders (like requirejs) support supplying an error callback as the third argument. Capturing module-load errors would be required for spec compliance.